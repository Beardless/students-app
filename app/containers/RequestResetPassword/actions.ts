import {
  REQUEST_RESET_PASSWORD,
  REQUEST_RESET_PASSWORD_ERROR,
  REQUEST_RESET_PASSWORD_SUCCESS
} from "./constants";

export function requestResetPasswordAction(
  payload: string
): { type: string; payload: string } {
  return {
    type: REQUEST_RESET_PASSWORD,
    payload
  };
}

export function requestResetPasswordActionSuccess(
  email: string
): { type: string; payload: string } {
  return {
    type: REQUEST_RESET_PASSWORD_SUCCESS,
    payload: email
  };
}

export function requestResetPasswordActionError(
  error: string
): { type: string; error: string } {
  return {
    type: REQUEST_RESET_PASSWORD_ERROR,
    error
  };
}
