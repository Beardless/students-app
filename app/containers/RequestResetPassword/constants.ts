export const REQUEST_RESET_PASSWORD: string =
  "courses/RequestResetPassword/REQUEST_RESET_PASSWORD";
export const REQUEST_RESET_PASSWORD_SUCCESS: string =
  "courses/RequestResetPassword/REQUEST_RESET_PASSWORD_SUCCESS";
export const REQUEST_RESET_PASSWORD_ERROR: string =
  "courses/RequestResetPassword/REQUEST_RESET_PASSWORD_ERROR";
