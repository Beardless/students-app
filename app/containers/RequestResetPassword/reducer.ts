import { fromJS } from "immutable";
import {
  REQUEST_RESET_PASSWORD,
  REQUEST_RESET_PASSWORD_ERROR,
  REQUEST_RESET_PASSWORD_SUCCESS
} from "./constants";

const initialState = fromJS({
  loading: false,
  error: false,
  success: false,
  email: null
});

function requestResetPassword(state = initialState, action) {
  switch (action.type) {
    case REQUEST_RESET_PASSWORD:
      return state.set("loading", true).set("error", false);
    case REQUEST_RESET_PASSWORD_SUCCESS:
      return state
        .set("loading", false)
        .set("success", true)
        .set("email", action.payload);
    case REQUEST_RESET_PASSWORD_ERROR:
      return state.set("loading", false).set("error", action.error);
    default:
      return state;
  }
}

export default requestResetPassword;
