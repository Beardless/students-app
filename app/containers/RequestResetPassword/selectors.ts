/**
 * Auth selectors
 */

import { createSelector } from "reselect";

const selectRequestResetPassword = state => state.get("requestResetPassword");
const makeSelectRequestResetPasswordLoading = () =>
  createSelector(
    selectRequestResetPassword,
    requestResetPasswordState => requestResetPasswordState.get("loading")
  );
const makeSelectRequestResetPasswordError = () =>
  createSelector(
    selectRequestResetPassword,
    requestResetPasswordState => requestResetPasswordState.get("error")
  );
const makeSelectRequestResetPasswordSuccess = () =>
  createSelector(
    selectRequestResetPassword,
    requestResetPasswordState => requestResetPasswordState.get("success")
  );
const makeSelectRequestResetPasswordEmail = () =>
  createSelector(
    selectRequestResetPassword,
    requestResetPasswordState => requestResetPasswordState.get("email")
  );

export {
  selectRequestResetPassword,
  makeSelectRequestResetPasswordLoading,
  makeSelectRequestResetPasswordError,
  makeSelectRequestResetPasswordSuccess,
  makeSelectRequestResetPasswordEmail
};
