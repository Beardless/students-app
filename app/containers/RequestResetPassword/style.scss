.container {
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;

  .success {
    max-width: 320px;
    padding: 16px;
    color: #48c55c;
    border-radius: 4px;
    background-color: rgba(72, 197, 92, 0.12);
  }
}
