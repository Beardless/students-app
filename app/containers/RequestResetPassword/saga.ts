import { call, put, takeEvery } from "redux-saga/effects";
import request from "../../utils/request";
import {
  requestResetPasswordActionError,
  requestResetPasswordActionSuccess
} from "./actions";
import { REQUEST_RESET_PASSWORD } from "./constants";
import { getAuthorizationHeaders } from "../../utils/userUtils";

function* requestResetPassword({ payload }: { payload: string }) {
  try {
    const options = {
      method: "POST",
      url: `/users/password/reset_request`,
      data: JSON.stringify({ email: payload }),
      headers: getAuthorizationHeaders()
    };

    console.log(options);

    yield call(request, options);

    yield put(requestResetPasswordActionSuccess(payload));
  } catch (err) {
    const { message } = err.response.data;
    yield put(requestResetPasswordActionError(message));
  }
}

export default function* requestResetPasswordFlow() {
  // @ts-ignore
  yield takeEvery(REQUEST_RESET_PASSWORD, requestResetPassword);
}
