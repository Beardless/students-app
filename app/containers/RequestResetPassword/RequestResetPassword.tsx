import * as React from "react";
import Fields from "../../components/Fields";
import i18n from "i18next";
import { AuthFieldsInterface } from "../Auth/interfaces";
import "./style.scss";
import { getValueFromFields, updateFields } from "../../utils/fields-logic";

interface RequestResetPasswordState {
  fields: AuthFieldsInterface;
}

interface RequestResetPasswordProps {
  success: boolean;
  error: boolean | string;
  loading: boolean;
  email: string | null;
  onSubmit: (payload: string) => void;
}

class RequestResetPassword extends React.Component<
  RequestResetPasswordProps,
  RequestResetPasswordState
> {
  constructor(props) {
    super(props);
    this.state = {
      fields: [
        {
          type: "login",
          label: i18n.t("Login"),
          value: "",
          disabled: false,
          error: null
        }
      ]
    };
  }

  handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const { fields } = this.state;
    this.props.onSubmit(getValueFromFields(fields, "login"));
  };

  handleUserInput = (fieldName: string, value: string) => {
    const updatedFields = updateFields(fieldName, value, this.state.fields);

    this.setState({ fields: updatedFields });
  };

  render() {
    const { fields } = this.state;
    const { error, loading, success, email } = this.props;

    if (success) {
      return (
        <div className={"container"}>
          <div className={"success"}>
            {i18n.t("An e-mail has been sent to", { value: email })}
          </div>
        </div>
      );
    }

    return (
      <div className={"container"}>
        <Fields
          subtitle={i18n.t("Provide your email")}
          label={i18n.t("You forgot a password")}
          buttonLabel={i18n.t("Submit")}
          fields={fields}
          isLoading={loading}
          error={error}
          onUserInput={this.handleUserInput}
          onSubmit={this.handleSubmit}
        />
      </div>
    );
  }
}

export default RequestResetPassword;
