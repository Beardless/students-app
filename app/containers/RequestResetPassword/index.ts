import RequestResetPassword from "./RequestResetPassword";
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import { requestResetPasswordAction } from "./actions";
import {
  makeSelectRequestResetPasswordEmail,
  makeSelectRequestResetPasswordError,
  makeSelectRequestResetPasswordLoading,
  makeSelectRequestResetPasswordSuccess
} from "./selectors";
import { compose } from "redux";
import reducer from "./reducer";
import saga from "./saga";
import injectSaga from "../../utils/injectSaga";
import injectReducer from "../../utils/injectReducer";

const mapDispatchToProps = dispatch => ({
  onSubmit: (payload: string) => dispatch(requestResetPasswordAction(payload))
});

const mapStateToProps = createStructuredSelector({
  loading: makeSelectRequestResetPasswordLoading(),
  error: makeSelectRequestResetPasswordError(),
  success: makeSelectRequestResetPasswordSuccess(),
  email: makeSelectRequestResetPasswordEmail()
});

const withRequestResetPasswordReducer = injectReducer({
  key: "requestResetPassword",
  reducer
});
const withRequestResetPasswordSaga = injectSaga({
  key: "requestResetPassword",
  saga
});

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps
);

export default compose<any>(
  withRequestResetPasswordReducer,
  withRequestResetPasswordSaga,
  withConnect
)(RequestResetPassword);
