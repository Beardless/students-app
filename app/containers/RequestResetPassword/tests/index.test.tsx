import * as React from "react";
import { shallow } from "enzyme";
import { mapDispatchToProps } from "../index";
import { requestResetPasswordAction } from "../actions";
import RequestResetPassword from "../RequestResetPassword";

describe("<RequestResetPassword />", () => {
  it("should render RequestResetPassword component", () => {
    const renderedComponent = shallow(
      <RequestResetPassword
        loading={false}
        success={false}
        error={false}
        email={null}
        onSubmit={() => {}}
      />
    );

    expect(renderedComponent).toMatchSnapshot();
  });

  describe("mapDispatchToProps", () => {
    describe("onSubmit", () => {
      it("should be injected", function() {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        expect(result.onSubmit).toBeDefined();
      });

      it("should dispatch requestResetPasswordAction when called", () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        const payload = "randomString";
        result.onSubmit(payload);
        expect(dispatch).toHaveBeenCalledWith(
          requestResetPasswordAction(payload)
        );
      });
    });
  });
});
