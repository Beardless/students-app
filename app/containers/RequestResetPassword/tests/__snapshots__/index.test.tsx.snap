// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestResetPassword /> should render RequestResetPassword component 1`] = `
<div
  className="container"
>
  <Fields
    error={false}
    fields={
      Array [
        Object {
          "disabled": false,
          "error": null,
          "label": undefined,
          "type": "login",
          "value": "",
        },
      ]
    }
    isLoading={false}
    onSubmit={[Function]}
    onUserInput={[Function]}
  />
</div>
`;
